### PR TITLE
filter all code over `pyupgrade --py38-plus`

### PR DESCRIPTION
**Pull request by kloczek**

### CODE REVIEW

**Review by Pierre-Sassoulas — COMMENTED** [review]
I'm wondering it we should keep pyupgrade in the pre-commit hooks then. ruff was at feature parity at some point but not anymore it seems. Or we could wait for ruff to keep up with feature parity and close this MR until then.

**Comment by The-Compiler** [comment]
This might just be a bug in ruff, with some special case (maybe the `**locals()`?) that made it miss this particular case?

**Comment by kloczek** [comment]
> This might just be a bug in ruff, with some special case (maybe the `**locals()`?) that made it miss this particular case?

`ruff` still is not able to upgrade the code between python version so IMO it make sense to have both in pre-commit.

BTW soon in Oct this year python 3.8 will be EOSed.
Currently `pyupgrade --py39-plus` causes some pytest self test issues.
It wold be good if someone will try to review those generated changes to adapt pytest code and/or report some issues in `pyupgrade`.
Even more units are failing with `pyupgrade --py3{9,10,11,12}-plus` as well.
Just in case .. it is not critical but more ItWouldBeGoodToHave™️ ..

**Comment by The-Compiler** [comment]
> `ruff` still is not able to upgrade the code between python version so IMO it make sense to have both in pre-commit.

ruff has [integrated pyupgrade functionality](https://docs.astral.sh/ruff/rules/#pyupgrade-up) for a long time already (possibly since the beginning?).

**Comment by kloczek** [comment]
> ruff has [integrated pyupgrade functionality](https://docs.astral.sh/ruff/rules/#pyupgrade-up) for a long time already (possibly since the beginning?).

Just try to take few modules and filter over `ruff` and than over 'pyupgrade`.
I understand that this was intention of the `ruff`. Reality is that `pyupgrade` is able to generate much more changes.

**Comment by The-Compiler** [comment]
You'll need to enable the respective rules (like pytest does).

**Comment by kloczek** [comment]
> You'll need to enable the respective rules (like pytest does).

And I've tested that with those rules several times.
`ruff` still is ot able to filter the same way code as `pyupgrade`.
Really .. please try test that.

**Comment by Pierre-Sassoulas** [comment]
We activated all of ruff's pyupgrade messages:

https://github.com/pytest-dev/pytest/blob/0070151c630b30e38381f1a3a00e263de4665e26/pyproject.toml#L110

So yes, not feature parity, but the diff of this MR is the only change pyupgrade would have applied in 4 months. (https://github.com/pytest-dev/pytest/pull/11911 merge on 2024-02-02). I'm not sure it's worth the extra wait for each commit.

Applying pyupgrade manually when each python interpreter dropped is also a possibility. That way we don't have to wait longer for each git commit and we can still benefit from pyupgrade's features not included in ruff, eventually.

**Comment by nicoddemus** [comment]
>Applying pyupgrade manually when each python interpreter dropped is also a possibility. That way we don't have to wait longer for each git commit and we can still benefit from pyupgrade's features not included in ruff, eventually.

I'm OK with this, that's what I do in my other projects: I apply `pyupgrade` as part of the process of dropping a supported Python version. Of course, in pytest this would open the possibility of a "obsolete" pattern being re-introduced, but I think this is rare/not really  a problem in general.

**Comment by RonnyPfannschmidt** [comment]
We could also just report findings upstream so that by the time we drop,ruff has parity

**Comment by Pierre-Sassoulas** [comment]
I pushed what I think is a good compromise: added pyupgrade back as a manual stage, this make it easily launchable with ``pre-commit run --hook-stage manual pyupgrade -a`` when we drop an interpreter (or anytime really) in 8f61d5d5723f4304bcf59e7c2639224095b31089.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -43,6 +43,11 @@ repos:
     - id: pyproject-fmt
       # https://pyproject-fmt.readthedocs.io/en/latest/#calculating-max-supported-python-version
       additional_dependencies: ["tox>=4.9"]
+-   repo: https://github.com/asottile/pyupgrade
+    rev: v3.15.2
+    hooks:
+    -   id: pyupgrade
+        stages: [manual]
```
Comment (nicoddemus) [inline .pre-commit-config.yaml:50]: TIL

**Review by nicoddemus — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Pierre-Sassoulas) [reply .pre-commit-config.yaml:50]: Previously added pylint this way so can be launched similarly with ``pre-commit run --hook-stage manual pylint -a`` :)

**Comment by kloczek** [comment]
Thank you 👍

**Pull request merged**